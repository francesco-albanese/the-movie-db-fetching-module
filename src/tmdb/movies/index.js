import { tmdbClient, urls } from '#tmdb/tmdb.const'
import { transformAllMoviesResponse, transformMovieDetailsResponse } from '#transformation'

import { parseStringifiedData } from '#utils'

export const getMovies = async(category, language) => {
  
  try {
    const { data }  = await tmdbClient.get(urls[ category ], {
      params: {
        language
      },
      transformResponse: [
        data => parseStringifiedData(data),
        ({ results }) => transformAllMoviesResponse(results)
      ]
    })

    return data
  }

  catch (e) {
    throw e
  }
}

export const getMovieById = async(movie_id, language) => {
  
  try {
    const { data }  = await tmdbClient.get(urls.movie(movie_id), {
      params: {
        language
      },
      transformResponse: [
        data => parseStringifiedData(data),
        data => transformMovieDetailsResponse(data)
      ]
    })

    return data
  }

  catch (e) {
    throw e
  }
}