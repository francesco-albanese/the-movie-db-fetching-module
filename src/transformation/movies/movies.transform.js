import { isEmpty } from 'lodash-es'
import { getMovieImage } from '#utils'

export const transformAllMoviesResponse = (data = []) => {

  if (isEmpty(data)) {
    return []
  } 
  
  return data.map(movie => {

    const { 
      genre_ids,
      id,
      poster_path,
      title
    } = movie

    const image = isEmpty(poster_path) 
      ? 'https://via.placeholder.com/411x617.png?text=No+Image+Available'
      : getMovieImage(poster_path)

    return {
      genre_ids,
      id,
      image,
      title
    }

  })
}