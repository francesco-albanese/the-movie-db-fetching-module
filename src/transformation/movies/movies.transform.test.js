import { transformAllMoviesResponse } from './movies.transform'
import * as Fixtures from './__fixtures__/movies.transform.fixtures'

describe('transformAllMoviesResponse', () => {
  it('returns an empty array if data is missing', () => {
    expect(transformAllMoviesResponse()).toEqual([])
  })

  it('returns transformed movies', () => {
    const result = transformAllMoviesResponse(Fixtures.data)
    expect(result).toMatchSnapshot()
  })
})