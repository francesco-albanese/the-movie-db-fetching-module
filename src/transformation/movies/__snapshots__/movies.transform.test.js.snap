// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformAllMoviesResponse returns transformed movies 1`] = `
Array [
  Object {
    "description": "When Eddie Brock acquires the powers of a symbiote, he will have to release his alter-ego \\"Venom\\" to...",
    "genre_ids": Array [
      878,
    ],
    "id": 335983,
    "image": "undefined/2uNW4WbgBXL25BAbXGLnLqX71Sw.jpg",
    "title": "Venom",
    "vote_average": 6.6,
  },
  Object {
    "description": "Laurie Strode comes to her final confrontation with Michael Myers, the masked figure who has haunted...",
    "genre_ids": Array [
      27,
      53,
    ],
    "id": 424139,
    "image": "undefined/bXs0zkv2iGVViZEy78teg2ycDBm.jpg",
    "title": "Halloween",
    "vote_average": 6.7,
  },
  Object {
    "description": "Elastigirl springs into action to save the day, while Mr. Incredible faces his greatest challenge ye...",
    "genre_ids": Array [
      28,
      12,
      16,
      10751,
    ],
    "id": 260513,
    "image": "undefined/x1txcDXkcM65gl7w20PwYSxAYah.jpg",
    "title": "Incredibles 2",
    "vote_average": 7.6,
  },
  Object {
    "description": "Traveling in search of the rare ingredient, “sky fish”  Meliodas and Hawk arrive at a palace that fl...",
    "genre_ids": Array [
      28,
      12,
      14,
      16,
    ],
    "id": 507569,
    "image": "undefined/lgYGPujZNckyKwQQNwMLmkWoKui.jpg",
    "title": "The Seven Deadly Sins: Prisoners of the Sky",
    "vote_average": 6,
  },
  Object {
    "description": "Seasoned musician Jackson Maine discovers—and falls in love with—struggling artist Ally. She has jus...",
    "genre_ids": Array [
      18,
      10402,
      10749,
    ],
    "id": 332562,
    "image": "undefined/wrFpXMNBRj2PBiN4Z5kix51XaIZ.jpg",
    "title": "A Star Is Born",
    "vote_average": 7.6,
  },
  Object {
    "description": "Robert McCall, who serves an unflinching justice for the exploited and oppressed, embarks on a relen...",
    "genre_ids": Array [
      53,
      28,
      80,
    ],
    "id": 345887,
    "image": "undefined/cQvc9N6JiMVKqol3wcYrGshsIdZ.jpg",
    "title": "The Equalizer 2",
    "vote_average": 6.2,
  },
  Object {
    "description": "Martial arts comedy following a group of kung fu legends banding together to take on the bad guys. T...",
    "genre_ids": Array [
      28,
      35,
    ],
    "id": 523873,
    "image": "undefined/rW0A73hjzPWVwADlCTLnjLhAFLX.jpg",
    "title": "Kung Fu League",
    "vote_average": 4,
  },
  Object {
    "description": "When a young nun at a cloistered abbey in Romania takes her own life, a priest with a haunted past a...",
    "genre_ids": Array [
      27,
      9648,
      53,
    ],
    "id": 439079,
    "image": "undefined/sFC1ElvoKGdHJIWRpNB3xWJ9lJA.jpg",
    "title": "The Nun",
    "vote_average": 5.8,
  },
  Object {
    "description": "An elite group of American operatives, aided by a top-secret tactical command team, must transport a...",
    "genre_ids": Array [
      28,
      53,
    ],
    "id": 347375,
    "image": "undefined/2L8ehd95eSW9x7KINYtZmRkAlrZ.jpg",
    "title": "Mile 22",
    "vote_average": 5.8,
  },
  Object {
    "description": "A deep sea submersible pilot revisits his past fears in the Mariana Trench, and accidentally unleash...",
    "genre_ids": Array [
      28,
      878,
      53,
    ],
    "id": 345940,
    "image": "undefined/eyWICPcxOuTcDDDbTMOZawoOn8d.jpg",
    "title": "The Meg",
    "vote_average": 6,
  },
  Object {
    "description": "Singer Freddie Mercury, guitarist Brian May, drummer Roger Taylor and bass guitarist John Deacon tak...",
    "genre_ids": Array [
      18,
      10402,
    ],
    "id": 424694,
    "image": "undefined/jf7jiy0rDVn1ljUwv3pWKousQ1p.jpg",
    "title": "Bohemian Rhapsody",
    "vote_average": 8.5,
  },
  Object {
    "description": "Teddy Walker is a successful salesman whose life takes an unexpected turn when he accidentally blows...",
    "genre_ids": Array [
      35,
    ],
    "id": 454293,
    "image": "undefined/1NSMAaBPSbWv7sGmF8oLGMNb8Qm.jpg",
    "title": "Night School",
    "vote_average": 6.3,
  },
  Object {
    "description": "As the Avengers and their allies have continued to protect the world from threats too large for any ...",
    "genre_ids": Array [
      12,
      878,
      28,
      14,
    ],
    "id": 299536,
    "image": "undefined/7WsyChQLEftFiDOVTGkv3hFpyyt.jpg",
    "title": "Avengers: Infinity War",
    "vote_average": 8.3,
  },
  Object {
    "description": "A look at the life of the astronaut, Neil Armstrong, and the legendary space mission that led him to...",
    "genre_ids": Array [
      36,
      18,
    ],
    "id": 369972,
    "image": "undefined/i91mfvFcPPlaegcbOyjGgiWfZzh.jpg",
    "title": "First Man",
    "vote_average": 7.2,
  },
  Object {
    "description": "From the outer reaches of space to the small-town streets of suburbia, the hunt comes home. Now, the...",
    "genre_ids": Array [
      27,
      878,
      28,
      53,
    ],
    "id": 346910,
    "image": "undefined/wMq9kQXTeQCHUZOG4fAe5cAxyUA.jpg",
    "title": "The Predator",
    "vote_average": 5.3,
  },
  Object {
    "description": "Just when his time under house arrest is about to end, Scott Lang puts again his freedom at risk to ...",
    "genre_ids": Array [
      28,
      12,
      878,
      35,
      10751,
    ],
    "id": 363088,
    "image": "undefined/rv1AWImgx386ULjcf62VYaW8zSt.jpg",
    "title": "Ant-Man and the Wasp",
    "vote_average": 6.9,
  },
  Object {
    "description": "An untested American submarine captain teams with U.S. Navy Seals to rescue the Russian president, w...",
    "genre_ids": Array [
      28,
      53,
    ],
    "id": 399402,
    "image": "undefined/82eQFmaLjRv3othN6BV3Ac80Abh.jpg",
    "title": "Hunter Killer",
    "vote_average": 5.8,
  },
  Object {
    "description": "Framed and on the run, a former FBI agent must save his family from a blazing fire in the world's ta...",
    "genre_ids": Array [
      28,
    ],
    "id": 447200,
    "image": "undefined/5LYSsOPzuP13201qSzMjNxi8FxN.jpg",
    "title": "Skyscraper",
    "vote_average": 6.1,
  },
  Object {
    "description": "After a hunting expedition goes awry, a young caveman struggles against the elements to find his way...",
    "genre_ids": Array [
      12,
      18,
    ],
    "id": 399360,
    "image": "undefined/afdZAIcAQscziqVtsEoh2PwsYTW.jpg",
    "title": "Alpha",
    "vote_average": 5.4,
  },
  Object {
    "description": "Working-class family man Christopher Robin encounters his childhood friend Winnie-the-Pooh, who help...",
    "genre_ids": Array [
      16,
      12,
      35,
      10751,
      14,
    ],
    "id": 420814,
    "image": "undefined/xR5w0he6czZkcAz459a4iPBqXGe.jpg",
    "title": "Christopher Robin",
    "vote_average": 7.5,
  },
]
`;
